test/aws.test.js

```diff
@@ -0,0 +1,98 @@
+/* global describe it before after */
+
+const assert = require('assert')
+const kmsHelper = require('../lib/kms-helper')
+const AWS = require('aws-sdk-mock')
+const sinon = require('sinon')
+const lambdaTester = require('lambda-tester')
+const handler = require('../index.js').handler
+const event = require('../event.json')
+const KinesisWriter = require('../lib/KinesisWriter')
+
+const getRDSString = (file) => {
+  return require('dotenv').config({ path: `${file}` }).parsed['DISCOVERY_STORE_CONNECTION_URI']
+}
+
+describe('AWS', function () {
+  this.timeout(5000)
+  require('dotenv').config({ path: './.env' })
+
+  describe('kmsHelper util', () => {
+    it('should fail when there is no encrypted key', () => {
+      process.env['DISCOVERY_STORE_CONNECTION_URI'] = ''
+      try {
+        kmsHelper.decryptDbCreds()
+      } catch (e) {
+        assert.equal(e.message, 'Missing DISCOVERY_STORE_CONNECTION_URI env variable; aborting.')
+      }
+    })
+
+    it('should call decrypt the db uri', () => {
+      process.env['DISCOVERY_STORE_CONNECTION_URI'] = getRDSString('./test.deploy.env')
+      return kmsHelper.decryptDbCreds()
+        .then(value => {
+          assert.equal(value, 'postgresql://localhost:5432/discovery')
+        })
+    })
+  })
+
+  describe('Connecting to KMS for DB URI string', () => {
+    var spyDecryptFunc
+
+    before(() => {
+      process.env['DISCOVERY_STORE_CONNECTION_URI'] = getRDSString('./deploy.env')
+      spyDecryptFunc = sinon.spy()
+      AWS.mock('KMS', 'decrypt', spyDecryptFunc)
+    })
+
+    it('should call kms.decrypt', () => {
+      kmsHelper.decryptDbCreds()
+      assert.equal(spyDecryptFunc.calledOnce, true)
+    })
+
+    after(() => AWS.restore('KMS', 'decrypt'))
+  })
+
+  describe('Connecting to Kinesis', () => {
+    require('dotenv').config({ path: './deploy.env' })
+    require('dotenv').config({ path: './.env' })
+    var spyPutRecordsFunc
+
+    before(() => {
+      spyPutRecordsFunc = sinon.stub()
```

Comment:
Here's one issue. Should this be...

```js
spyPutRecordsFunc = sinon.spy((params, callback) => callback(null, `mock called with ${params}`))
```

The problem seems to be that AWS.mock doesn't actually mock the function in a way that resembles the real call. My code depends on AWS.Kinesis().putRecord accepting (and firing) a callback as second param. `sinon.spy()` creates a function that tracks when and how many times it's called, but doesn't do anything with the parameters it's given. Hence it doesn't know to execute the `callback` param. Hence I never call either `resolve` or `reject` in the callback I've passed to putRecords.

Per http://sinonjs.org/releases/v2.1.0/spies/ , one solution is to create the spy function as a wrapper around a function that actually calls a callback, as above.

Another issue seems to be the order in which AWS.Kinesis() is called: "NB: The AWS Service needs to be initialised inside the function being tested in order for the SDK method to be mocked " (https://github.com/dwyl/aws-sdk-mock ). It may need to be called inside the exported function (i.e. after AWS.mock has been called) Encountering other aws issues atm so will resume tomorrow.